elasticsearch/src/main/java/com/hazelcast/jet/elasticsearch/ElasticSearchSinks.java

```diff
@@ -0,0 +1,160 @@
+     */
+    public static <T> Sink<T> elasticSearch(String name,
+                                            SupplierEx<RestHighLevelClient> clientSupplier,
+                                            SupplierEx<BulkRequest> bulkSupplier,
+                                            FunctionEx<T, IndexRequest> indexFn
+    ) {
+        return SinkBuilder
+                .sinkBuilder("elasticSearch-" + name,
+                        ctx -> new BulkContext(clientSupplier.get(), bulkSupplier))
+                .<T>receiveFn((bulkContext, item) -> bulkContext.add(indexFn.apply(item)))
+                .flushFn(BulkContext::bulk)
+                .destroyFn(BulkContext::close)
+                .build();
+    }
+
+    /**
+     * Creates a sink which indexes objects into the specified elastic-search
+     * using bulk requests.
+     * Rest client is configured with basic authentication.
+     *
+     * @param name         Name of the created sink
+     * @param bulkSupplier Bulk request supplier
+     * @param indexFn      Creates an {@link IndexRequest} for each object
+     */
+    public static <T> Sink<T> elasticSearch(String name,
+                                            String username, String password,
+                                            String hostname, int port,
+                                            SupplierEx<BulkRequest> bulkSupplier,
+                                            FunctionEx<T, IndexRequest> indexFn
+    ) {
+        return elasticSearch(name, () -> buildClient(username, password, hostname, port), bulkSupplier, indexFn);
+    }
+
+    private static RestHighLevelClient buildClient(String username, String password, String hostname, int port) {
+        CredentialsProvider credentialsProvider = new BasicCredentialsProvider();
+        credentialsProvider.setCredentials(ANY, new UsernamePasswordCredentials(username, password));
+        return new RestHighLevelClient(
+                RestClient.builder(new HttpHost(hostname, port))
+                          .setHttpClientConfigCallback(httpClientBuilder ->
+                                  httpClientBuilder.setDefaultCredentialsProvider(credentialsProvider))
+        );
+    }
+
+    private static final class BulkContext {
+
+        private final RestHighLevelClient client;
+        private final SupplierEx<BulkRequest> bulkSupplier;
+
+        private BulkRequest bulkRequest;
+
+        private BulkContext(RestHighLevelClient client, SupplierEx<BulkRequest> bulkSupplier) {
+            this.client = client;
+            this.bulkSupplier = bulkSupplier;
+            this.bulkRequest = bulkSupplier.get();
+        }
+
+        private void add(IndexRequest request) {
+            bulkRequest.add(request);
+        }
+
+        private void bulk() throws IOException {
+            client.bulk(bulkRequest, RequestOptions.DEFAULT);
+            bulkRequest = bulkSupplier.get();
```

Comment:
I don't quite understand this line. Why do we re-setting it's value to the user provided bulkrequest ?

Reply:
we want the user to create a new `BulkRequest` after sending the previous one